src/server.rs

```diff
@@ -1,50 +1,83 @@
-    println!("Server live!");
+/// Runs the server. The server is a long-running process that listens for messages and echoes
+/// them back the client.
+pub(crate) async fn run_server(opts: crate::Opts) {
+    // If a server address & port are provided as command-line inputs, use those, else use the
+    // default.
+    let server_address = opts
+        .address
+        .unwrap_or_else(|| ipv4_resolve(DEFAULT_SERVER_ADDRESS).unwrap());
 
+    println!("Starting server on {:?}", server_address);
+
+    // Bind a server-side socket to requested address and port. If "0" was provided as the port, the
+    // OS will allocate a port and the actual port used will be available in `actual_server_addr`.
+    //
+    // `outbound` is a `UdpSink`, we use it to send messages. `inbound` is `UdpStream`, we use it
+    // to receive messages.
+    //
+    // This is an async function, so we need to await it.
+    let (outbound, inbound, actual_server_addr) = bind_udp_bytes(server_address).await;
+
+    println!("Server is live! Listening on {:?}", actual_server_addr);
+
+    // The skeletal hydroflow spec for a server.
     let mut flow: Hydroflow = hydroflow_syntax! {
-        // Define shared inbound and outbound channels
-        inbound_chan = source_stream_serde(inbound) -> map(|udp_msg| udp_msg.unwrap()) -> tee();
-        outbound_chan = union() -> dest_sink_serde(outbound);
 
-        // Print all messages for debugging purposes
-        inbound_chan[1]
-            -> for_each(|(m, a): (Message, SocketAddr)| println!("{}: Got {:?} from {:?}", Utc::now(), m, a));
+        // Whenever a serialized message is received by the application from a particular address,
+        // a (serialized_payload, address_of_sender) pair is emitted by the `inbound` stream.
+        //
+        // `source_stream_serde` deserializes the payload into a
+        // (deserialized_payload, address_of_sender) pair.
+        inbound_chan = source_stream_serde(inbound) // `source_stream_serde` deserializes the payload
+            -> map(Result::unwrap); // If the deserialization was unsuccessful, this line will panic.
+
+        // Mirrors the inbound process on the outbound side.
+        // `dest_sink_serde` accepts a (`Message`, `SocketAddr`) pair and serializes the `Message`
+        // using `serde`, converting it to a (serialized_payload, address_of_receiver) pair.
+        // `outbound` transmits the serialized_payload to the address.
+        outbound_chan = union() -> dest_sink_serde(outbound);
 
         // Demux and destructure the inbound messages into separate streams
-        inbound_demuxed = inbound_chan[0]
-            ->  demux(|(msg, addr), var_args!(echo, heartbeat, errs)|
-                    match msg {
+        inbound_demuxed = inbound_chan
+            -> inspect(|(m, a): &(Message, SocketAddr)| println!("{}: Got {:?} from {:?}", Utc::now(), m, a)) // For debugging purposes.
```

Comment:
Thought this was a good place to introduce `inspect`. It removes the `tee` from the program, but this is OK because an [earlier chapter](https://hydro.run/docs/hydroflow/quickstart/example_6_unreachability) introduces it.